server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/StreamCompletionRunnable.java

```diff
@@ -0,0 +1,88 @@
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.blobstore.stream.read.listener;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.common.io.Channels;
+import org.opensearch.common.io.InputStreamContainer;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * StreamCompletionListener transfers the provided stream into the specified file path using a {@link FileChannel}
+ * instance. It performs offset based writes to the file and notifies the {@link FileCompletionListener} on completion.
+ *
+ * @opensearch.internal
+ */
+public class StreamCompletionRunnable implements Runnable {
+    private final InputStreamContainer blobPartStreamContainer;
+    private final Path fileLocation;
+    private final AtomicBoolean anyPartStreamFailed;
+    private final FileCompletionListener fileCompletionListener;
+    private static final Logger logger = LogManager.getLogger(StreamCompletionRunnable.class);
+
+    // 8 MB buffer for transfer
+    private static final int BUFFER_SIZE = 8 * 1024 * 2024;
+
+    public StreamCompletionRunnable(
+        InputStreamContainer blobPartStreamContainer,
+        Path fileLocation,
+        AtomicBoolean anyPartStreamFailed,
+        FileCompletionListener fileCompletionListener
+    ) {
+        this.blobPartStreamContainer = blobPartStreamContainer;
+        this.fileLocation = fileLocation;
+        this.anyPartStreamFailed = anyPartStreamFailed;
+        this.fileCompletionListener = fileCompletionListener;
+    }
+
+    @Override
+    public void run() {
+        // Ensures no writes to the file if any stream fails.
+        if (anyPartStreamFailed.get() == false) {
+            try (FileChannel outputFileChannel = FileChannel.open(fileLocation, StandardOpenOption.WRITE, StandardOpenOption.CREATE)) {
+                try (InputStream inputStream = blobPartStreamContainer.getInputStream()) {
+                    long streamOffset = blobPartStreamContainer.getOffset();
+                    final byte[] buffer = new byte[BUFFER_SIZE];
+                    int bytesRead;
+                    while ((bytesRead = inputStream.read(buffer)) != -1) {
+                        Channels.writeToChannel(buffer, 0, bytesRead, outputFileChannel, streamOffset);
+                        streamOffset += bytesRead;
+                    }
+                }
+            } catch (IOException e) {
+                processFailure(e);
+                return;
+            }
+            fileCompletionListener.onResponse(null);
+        }
+    }
+
+    void processFailure(Exception e) {
+        logger.error("Downloading part stream failed for file {}: {}", fileLocation, e);
```

Comment:
I think this should be removed because `e` gets passed on via the listener callback and the caller should have the responsibility to do whatever is appropriate with the exception.